crml/staking/src/mock.rs

```diff
@@ -605,26 +539,22 @@ pub(crate) fn bond_nominator(stash: AccountId, ctrl: AccountId, val: Balance, ta
 	assert_ok!(Staking::nominate(Origin::signed(ctrl), target));
 }
 
-pub fn advance_session() {
-	let current_index = Session::current_index();
-	start_session(current_index + 1);
-}
+pub fn start_session(index: SessionIndex) {
+	assert!(Session::current_index() <= index);
 
-pub fn start_session(session_index: SessionIndex) {
-	// Compensate for session delay
-	let session_index = session_index + 1;
```

Comment:
what was the reason for changing this? now there is lot's of house keeping in the tests, not sure if it was relevant or this PR

Reply:
Because it has wrongly coupled the concept of a block number to session number. The right way of changing to a particular session number is through rotating sessions. This was inconsistent with the tests that we have in the runtime modules as well. The session_index + 1 that was needed should have indicated that this wasn't correct.

Reply:
There wasn't a lot of housekeeping and I have already done that too. There are a few places where we need to start a block as well which is better to be done as a step in the test to make it clear what's happning.

Reply:
I think it's harder to understand the tests now, its easier if it always does the same thing but now it sometimes does session increment and sometimes block number increment and it's not clear why. it looks upstream like this now: https://github.com/paritytech/substrate/blob/9c5da5cc68e710d4d44b6bbeea6fb5a5e49b7134/frame/staking/src/mock.rs#L617-L628 which is easier to follow

Reply:
start_era and start_sessions together are used in about 100 times in our tests. However, there are only 12 places where we needed to start a block as well. The reason for starting a block can be mentioned as a comment in the test and I am happy to do that. In our staking module design, when an era starts, the rewards are enqueued. However, the rewards are only paid on subsequent blocks. Mixing these different functions (starting and authoring a block + starting a session) is not giving us the fine control of the environment that we need for tests. Starting a block affects the authorship as well, which we should do on clear intentions.
At the end of the day, start_era and start_session are helper functions inside the test. You can do lots of things in a function and name it whatever you want. However, best design practices recommend 1. Good appropriate names 2. [Single responsibility](https://en.wikipedia.org/wiki/Single-responsibility_principle)